[FLINK-4116] Metrics documentation

zentol opened the pull request.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+Metrics
+-------------------
+
+Flink exposes a metric system that allows users gather and expose metrics to external systems.
```
"allows users to gather and ..." or "allows user gathering and ..."

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+### Registering metrics
+
+You can access the metric system from any user function that extends [RichFunction]({{ site.baseurl }}/apis/common/index.html#rich-functions) by calling getRuntimeContext().getMetricGroup().
```
`getRuntimeContext().getMetricGroup` should be enclosed by back ticks.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+### Registering metrics
+
+You can access the metric system from any user function that extends [RichFunction]({{ site.baseurl }}/apis/common/index.html#rich-functions) by calling getRuntimeContext().getMetricGroup().
+This method returns a MetricGroup object on which you can create and register new metrics.
```
`MetricGroup` should have backticks.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+You can access the metric system from any user function that extends [RichFunction]({{ site.baseurl }}/apis/common/index.html#rich-functions) by calling getRuntimeContext().getMetricGroup().
+This method returns a MetricGroup object on which you can create and register new metrics.
+
+If you want to count the number of records your user function has received you could use a Counter like this:
```
`Counter` in backticks.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+  }
+}
+
+{% endhighlight %}
```
Maybe we should also demonstrate how to register a custom metric.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+### Metric types
+
+Flink supports Counters, Gauges and Histograms.
```
backticks or writing counters, gauges and histograms without a capital letter.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+Flink supports Counters, Gauges and Histograms.
+
+A Counter is used to count something. The current value can be in- or decremented using `inc([long n])` or `dec([long n])`.
```
I've never seen the notation `inc([long n])` in the Flink documentation. Maybe it's better to specify the parameterless and the parameter variant explicitly.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+Flink supports Counters, Gauges and Histograms.
+
+A Counter is used to count something. The current value can be in- or decremented using `inc([long n])` or `dec([long n])`.
+You can create and registers a Counter by calling `counter(String name)` on a MetricGroup. Alternatively, you can pass an instance of your own Counter implementation along with the name.
```
register without *s*

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+A Histogram measure the distribution of long values.
+You can register one by calling histogram(String name, Histogram histogram) on a MetricGroup.
+
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
```
I think it would be better to present the different metric types either as a descriptive list similar to the description environment of latex or by using sub headings. It could also be more descriptive to give for each metric type a short code snippet how to register and interacting with it. That would explain their interface which is important for the user.

tillrohrmann replied:
Maybe we could also say which module contains the `DropWizardHistogramWrapper`

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+### Scope
+
+Every registered metric has an automatically assigned scope which represent the entities it is tied to. By default a metric that is registered in a user function will be scoped to the Operator in which the function runs, the Task/Job it belongs to and the TaskManager/Host it is executed on. This is referred to as the "system scope".
```
which represents

tillrohrmann replied:
operator with small letter.

tillrohrmann replied:
task/job with small letter

tillrohrmann replied:
Same for task manager/host

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+Every registered metric has an automatically assigned scope which represent the entities it is tied to. By default a metric that is registered in a user function will be scoped to the Operator in which the function runs, the Task/Job it belongs to and the TaskManager/Host it is executed on. This is referred to as the "system scope".
+
+You can define an additonal "user scope" by calling the MetricGroup#addGroup((int/String) name) method.
```
explicit enumeration of the different `addGroup` methods is better imo.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+{% endhighlight %}
+
+The name under which a metric is exported is based on both scopes and the name passed in the `counter() call. The order is always \<system_scope>\<user_scope>\<name>.
```
counter has only one backtick.

tillrohrmann replied:
Don't we insert a scope separating character in between the system, user and name scope?

zentol replied:
kind of. i guess technically the delimiter is part of every group. I wanted to avoid adding `.` between the scope parts as it may lead to the misunderstanding that there are always only 2 dots in the name.

tillrohrmann replied:
We could put it into backticks.

zentol replied:
I don't understand how backticks solve the issue i mentioned.

tillrohrmann replied:
Hmm I think you're right, since the periods won't probably be rendered differently. But you could put the whole name in quotation marks. Then it should be clear.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+The system scope allows the reported name to contain contextual information like the name of job it was registered in without requiring the user to pass this information manually.
+
+How the system scope affects the reported name for a metric can be modified by setting the following keys in the flink-conf.yaml. 
+Each of these keys expect a format string that may contain constants (e.g. "taskmanager") and variables (e.g. "\<task_id>") which will be replaced at runtime.
```
The config key definition should also be mirrored to https://ci.apache.org/projects/flink/flink-docs-master/setup/config.html. Actually I think that the config keys should be moved from `MetricRegistry` to `ConfigConstants`.

rmetzger replied:
I agree that we should mirror the config keys to the configuration page as well + move the config keys.

uce replied:
I agree, but note that there are other components who also do it outside of `ConfigConstants`.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+There is no restriction on the order of variables.
+
+Which format is applied to a metric depends on which entities it was scoped to.
+The `metrics.scope.operator` format will be applied to all metrics that were scoped to an Operator.
```
operator with small letter

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+Which format is applied to a metric depends on which entities it was scoped to.
+The `metrics.scope.operator` format will be applied to all metrics that were scoped to an Operator.
+The `metrics.scope.tm.job` format will be applied to all metrics that were only scoped to a Job and a TaskManager.
```
job and task manager.

tillrohrmann replied:
Maybe it would be good to add the explanation of the different scopes directly to the enumeration of the config keys instead of listing them in this paragraph.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+The `metrics.scope.operator` format will be applied to all metrics that were scoped to an Operator.
+The `metrics.scope.tm.job` format will be applied to all metrics that were only scoped to a Job and a TaskManager.
+
+This allows you to define explicitly how metrics on every level should be reported.
```
Giving a small example could be helpful.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+- `metrics.reporter.interval`: The interval between reports.
+  - Example: 10 SECONDS)
+
+By default Flink uses JMX to expose metrics.
```
JMX should also get a sub heading.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+The port for JMX can be configured by setting the `metrics.jmx.port` key. This parameter expects either a single port
+or a port range, with the default being 9010-9025. Which port is used in the end is shown in the relevant Job-/TaskManager log.
+
+All non-JMXReporters are not part of the distribution and have to be added to the classpath manually. (usually by putting the jar into /lib)
```
parenthesis starts after the period

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+All non-JMXReporters are not part of the distribution and have to be added to the classpath manually. (usually by putting the jar into /lib)
+
+You can write your own Reporter by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
+If the Reporter should send out reports regularly you have to implement the Scheduled interface as well.
```
Reporter and Scheduled should be put into backticks.

tillrohrmann replied:
Maybe we should explain what `Scheduled` indicates.

zentol replied:
Doesn't 'If the Reporter should send out reports regularly...' indicate what `Scheduled` means?

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+- `dmax`
+- `tmax`
+- `ttl`
+- `addressingMode` (UNICAST/MULTICAST)
```
Explanation for all of the parameters required.

tillrohrmann replied:
Which flink module do I have to add for this reporter? Maybe give code snippet which you can easily add to your `pom.xml` file.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+#### Graphite (org.apache.flink.metrics.graphite.GraphiteReporter)
+- `host`
+- `port`
```
Explanation of parameters

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+#### StatsD (org.apache.flink.metrics.statsd.StatsDReporter)
+- `host`
+- `port`
```
Explanation of parameters

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+      </td>
+    </tr>
+  </tbody>
+</table>
```
Explanation of metrics could be helpful.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+#### StatsD (org.apache.flink.metrics.statsd.StatsDReporter)
+- `host`
+- `port`
+
```
We should also say how the different reporters expose certain metrics. E.g. the histogram metric reports the mean, median, percentiles etc. The names under which this happens might differ between different reporters.

tillrohrmann commented:
Good first draft of the metrics documentation @zentol. I think the documentation should be a bit more verbose containing more explanations and code examples. This will help the user to quickly get started and understanding the metrics system.

rmetzger commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,211 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+Flink supports the following systems with their respective configuration parameters:
+
+#### Ganglia (org.apcahe.flink.metrics.ganglia.GangliaReporter)
```
typo in apcahe

zentol commented:
I've addressed all comments, except the config page.

zentol commented:
also missing: specific reporter naming

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+### Scope
+
+Every registered metric has an automatically assigned scope which represents the entities it is tied to. By default a metric that is registered in a user function will be scoped to the operator in which the function runs, the task/job it belongs to and the taskManager/host it is executed on. This is referred to as the "system scope".
```
taskManager -> task manager

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+Every registered metric has an automatically assigned scope which represents the entities it is tied to. By default a metric that is registered in a user function will be scoped to the operator in which the function runs, the task/job it belongs to and the taskManager/host it is executed on. This is referred to as the "system scope".
+
+You can define an additonal "user scope" by calling the either `MetricGroup#addGroup(String name)` or `MetricGroup#addGroup(int name)`.
```
"the" before "either" can be removed.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+If you also want to include the task name, but omit the taskmanager information you can specify the following format:
+
+`metrics.scope.tm.operator: \<host>.\<job_name>.\<task_name>.\<operator_name>.\<subtask_index>`
```
Do we need the `\` in front of the variables? They are also shown in the rendered html document.

zentol replied:
Without the `\` they are interpreted as tags and not as text.

I used the build_docs.sh script to preview the docs, and the backslashes did not appear.

But I could (and probably should) replace all the < and > with `&lt;` and `&gt;` respectively.

tillrohrmann replied:
I used Chrome to look at the docs. Maybe it depends on the browser. Replacing them by `&lt;` and `&gt;` sounds like the right way to go.

uce replied:
Same for me in Safari

zentol replied:
oh, in this case we don't need them since they are in backticks.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+- `metrics.reporter.interval`: The interval between reports.
+  - Example: 10 SECONDS
+
+You can write your own Reporter by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
```
Either putting `Reporter` in backticks or using a small letter would be better.

tillrohrmann commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+  - Example: 10 SECONDS
+
+You can write your own Reporter by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
+If the Reporter should send out reports regularly you have to implement the Scheduled interface as well.
```
Same applies to `Scheduled`

tillrohrmann commented:
The documentation has reached a really good state. I had only some minor comments left.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
 
 {% top %}
+
+Metrics
```
I would make this a separate page under `Programming Guide`. It's guide hidden under the `Basic API` concepts although I understand that it it is technically the right place for it.

zentol replied:
It's also quite big and blows up the Basic API page. It takes up a quarter of the page alone.

I'll move it.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+#### Counter
+
+A `Counter` is used to count something. The current value can be in- or decremented using `inc()/inc(long n)` or `dec()/dec(long n)`.
+You can create and register a `Counter` by calling `counter(String name)` on a MetricGroup.
```
Orthogonal discussion, but the name `counter` to create a counter (and others respectively) is a style break imo compared the other classes which follow Java-bean style. Is that something we should change for the release? I'm definitely fine with keeping it the way it is, but it's something I've noticed.

zentol replied:
I'm not too happy with the names either.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+  @Override
+  public void open(Configuration config) {
+    // register a custom counter
+    this.counter = getRuntimeContext().getmetricGroup().counter("myCustomCounter", new CustomCounter());
```
the line breaks in my browser are bad, maybe move everything after `getMetricGroup()` to the next line? Furthermore, `getMetricGroup` has a lower letter `m`.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+#### Gauge
+
+A `Gauge` provides a value of any type on demand. In order to use a `Gauge` you must first create a class that implements the `org.apache.flink.metrics.Gauge` interface.
+There is not restriction for the type of the returned value.
```
btw does it not need to be serializable?

zentol replied:
no, reporters generally use toString() on the value returned. that should probably be documented though.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+  @Override
+  public void open(Configuration config) {
+    // register the gauge
+    getRuntimeContext().getmetricGroup().gauge("MyGauge", new Gauge<Integer>() {
```
`getmetricGroup` with lower letter `m`

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+      public Integer getValue() {
+        return valueToExpose;
+      }});
+    ...
```
My personal taste, but I think it's better to just remove this (and other) three dots from the examples. Makes copy pasting easier and it should be clear from context that other code can be added as needed.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+#### Histogram
+
+A Histogram measure the distribution of long values.
```
measureS with s

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+  @Override
+  public void open(Configuration config) {
+    // create and register a counter
+    this.histogram = getRuntimeContext().getMetricGroup().histogram("myHistogram", new MyHistogram());
```
same line break issue as the custom counter

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+#### Histogram
+
+A Histogram measure the distribution of long values.
+You can register one by calling histogram(String name, Histogram histogram) on a MetricGroup.
```
missing code formating for `histogram(...)` (via wrap in backtick)

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+}
+{% endhighlight %}
+
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
```
I would rephrase this along the lines of "Flink does not provide a default implementation for `Histogram`, but offers a wrapper that allows usage of Dropwizard histograms (see LINK TO CASS ON GITHUB).

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+{% endhighlight %}
+
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
+This wrapper is contained in the `flink-metrics-dropwizard` module.
```
- Link to how a user can add this module
- I would also add an example for this actually (as everyone who wants to use Histogram will need this)

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+### Reporter
+
+Metrics can be exposed to an external system by configuring a reporter in the `conf/flink-conf.yaml`.
```
Either `in conf/flink-conf.yaml` or `in the ... file`

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+  - Example: 10 SECONDS
+
+You can write your own `Reporter` by implementing the `org.apache.flink.metrics.reporter.MetricReporter` interface.
+If the Reporter should send out reports regularly you have to implement the `Scheduled` interface as well.
```
Are there any reporters which are not scheduled?

zentol replied:
JMX for example is not scheduled. I could also imagine Reporters that are queried from the outside.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+If the Reporter should send out reports regularly you have to implement the `Scheduled` interface as well.
+
+By default Flink uses JMX to expose metrics.
+All non-JMXReporters are not part of the distribution and have to be added to the classpath manually, either by putting the jar into /lib
```
I think we also link to an explanation in another docs page for this

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+All non-JMXReporters are not part of the distribution and have to be added to the classpath manually, either by putting the jar into /lib
+or including it in the job jar.
+
+Flink supports the following systems:
```
I think the `:` and the following sub sections don't go well together. I think either removing the sentence or saying something like `The following sections list the supported reporters` is better

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+<dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-ganglia</artifactId>
+      <version>1.1-SNAPSHOT</version>
```
We need to use version variables instead of hard coded values here (and in other places)

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+
+This could create the name `localhost.MyJob.MySource_->_MyOperator.MyOperator.0`.
+
+The following is a list of all variables available:
```
I think it makes sense to make this a subsection so that we can directly link to it etc.

uce commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+Flink only provides an interface for Histograms, but offers a Wrapper that allows usage of Codahale/DropWizard Histograms. (org.apache.flink.dropwizard.metrics.DropWizardHistogramWrapper)
+This wrapper is contained in the `flink-metrics-dropwizard` module.
+
+### Scope
```
I was wondering whether the names are case sensitive or not?

zentol replied:
the variable names? yes

uce commented:
Good job Chesnay! Reads very well. I like the way the text and code examples are mixed. I had some minor inline comments.

My biggest concern (not reflected in the inline comments) is the `Scope` section, which is too dense in my opinion. Do you think we can improve it somehow? 

Maybe add sub sections `System Scope`, `User Scope`, `Examples`, `Variables` for starters. The `hierarchical orders` stuff I did not get.

zentol commented on docs/apis/common/index.md:
```diff
@@ -1350,3 +1350,397 @@ You may specify program arguments before the job is executed. The plan visualiza
+</dependency>
+{% endhighlight %}
+
+Parameters:
```
yes

zentol commented:
I've made the following changes:
- rewritte scope section
- removed ... from examples
- FIX ALL THE TYPOS
- better line breaking
- added dependency&example for DWHWrapper
- replaced hard-coded dependency versions

tillrohrmann commented on docs/apis/metrics.md:
```diff
@@ -0,0 +1,441 @@
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-ganglia</artifactId>
+      <version>{{site.version}}</version>
+</dependency>
```
I think it is not enough to simply include this dependency to your job in order to use the Ganglia reporter. Instead it should be put in the lib directory, because the jar is required when the cluster is started. It only works if the cluster is started for a single job and the user jars are contained in the system classloader.

I think we should document this properly.

zentol replied:
can we fix this somehow? I would prefer it if users were not **required** to juggle around jars.

zentol replied:
Note: we probably need this long-term anyway (for per-job reporters)

tillrohrmann replied:
At the moment I fear there is no way around this, since we start the reporters when the cluster is started.

zentol replied:
right, of course. I've adjusted the docs.

tillrohrmann commented on docs/apis/metrics.md:
```diff
@@ -0,0 +1,441 @@
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-metrics-graphite</artifactId>
+      <version>{{site.version}}</version>
+</dependency>
```
We should document that the user has to put the `flink-metrics-graphite-VERSION-jar-with-dependencies.jar` in the `lib` folder or make it differently accessible to the cluster when starting it.

tillrohrmann replied:
The same probably applies to the `GangliaReporter`.

tillrohrmann commented on docs/apis/metrics.md:
```diff
@@ -0,0 +1,441 @@
+
+Parameters:
+
+- `host` - the ganglia server host
```
We should highlight that this is the value you set in `gmond.conf` `udp_recv_channel.bind`. Otherwise the UPD packages won't arrive. That was actually the problem I had the whole time. I'm just not good with tooling...

zentol replied:
Note that it is not mandatory to set the bind property. If it is not set then generally host should be set to localhost.

But yes, if it set it to something but localhost then naturally you have to set it to that value, will include that in the docs.

tillrohrmann replied:
The `udp_recv_channel.bind` is set per default in the `gmond.conf` if you set up Ganglia.
